### PR TITLE
Mob 3730 file preview doesnt work in chat for messages with attachments

### PR DESCRIPTION
**Pull request by DavDo**
**Jira issue:**
https://glia.atlassian.net/browse/MOB-3730

> [!IMPORTANT]
> Sorry for the big PR, I tried to split it into smaller commits 1 - rename, 2 - refactoring, 3 - unit tests(currently working on), but `refactoring` still is bigger than expected, but sometimes it is the only way to do refactorings.

**What was solved?**
- I renamed FileAttachment LocalAttachment to distinguish clearly between local and remote attachments
- I removed `SendMessagePayload` local wrapper since we can get all the information from core sdks `SendMessagePayload`
- I made changes to the `Chat` logic to use local attachments with `Uri` for visitor-sent attachments instead of downloading and caching them locally
- I removed the `Unsent` message type because currently, we're adding all the visitor messages as a preview and then updating their status in chat

To align chat behavior with iOS
- Now tapping on the image preview or attachment view will always open the image preview or attachment view page, even in case of an error.
- In case of an error, a visitor now has to click the message row or attachment container(not preview) to resend the message.

I will also add a video later to demonstrate these changes.

**Release notes:**

 - [ ] Feature
 - [x] Ignore
 - [ ] Release notes (Is it clear from the description here?)
 - [ ] Migration guide (If changes are needed for integrator already using the SDK - what needs to be communicated? Add underneath please)

**Additional info:**

 - [ ] Is the feature sufficiently tested? All tests fixed? Necessary unit, acceptance, snapshots added? Check that at least new public classes & methods are covered with unit tests
 - [ ] Did you add logging beneficial for troubleshooting of customer issues?
 - [ ] **Did you add new logging?** We would like the logging between platforms to be similar. Refer to [**Logging from Android SDKs** → **Things to consider for newly added logs**](https://glia.atlassian.net/wiki/spaces/ENG/pages/3568861468/Logging+from+Android+SDKs#Things-to-consider-for-newly-added-logs) in Confluence for more information.

**Screenshots:**

### CODE REVIEW

File: widgetssdk/src/main/java/com/glia/widgets/chat/ChatManager.kt [hunk 1]
```diff
@@ -195,23 +204,84 @@ internal class ChatManager(
...
             Action.ChatRestored, Action.None -> state
-            is Action.MessageClicked -> mapMessageClicked(action.messageId, state)
-            is Action.MessageSent -> mapMessageSent(action.message, state)
-            is Action.MessageSendError -> addErrorMessage(action.message, state)
-            is Action.RemoveMessage -> removeMessage(action.messageId, state)
+            is Action.AttachmentPreviewAdded -> mapAttachmentPreviewAdded(action.attachments, action.payload, state)
+            is Action.MessagePreviewAdded -> mapMessagePreviewAdded(action.visitorChatItem, action.payload, state)
+            is Action.OnMessageSent -> mapMessageSent(action.message, state)
+            is Action.OnSendMessageError -> mapSendMessageFailed(action.messageId, state)
+            is Action.OnRetryClicked -> mapRetryClicked(action.messageId, state)
```
Comment (andrews-moc) [inline widgetssdk/src/main/java/com/glia/widgets/chat/ChatManager.kt:218]: `Action.RemoveMessage` is not needed anymore, do I get it right?

File: widgetssdk/src/main/java/com/glia/widgets/chat/ChatManager.kt [hunk 2]
```diff
@@ -195,23 +204,84 @@ internal class ChatManager(
...
+        val messageIndex = chatItems.indexOfLast { it.id == payload.messageId }
+
+        if (messageIndex != -1) {
+            chatItems[messageIndex] = (chatItems[messageIndex] as VisitorChatItem).withStatus(VisitorItemStatus.PREVIEW)
+        }
+
+        if (files.isEmpty()) return@apply
+
+        val firstFileIndex = if (messageIndex != -1)
+            messageIndex + 1
```
Comment (andrews-moc) [inline widgetssdk/src/main/java/com/glia/widgets/chat/ChatManager.kt:236]: Could you please add a comment why file index is a message index incremented by 1?

File: widgetssdk/src/main/java/com/glia/widgets/chat/ChatManager.kt [hunk 3]
```diff
@@ -195,23 +204,84 @@ internal class ChatManager(
...
+        else
+            chatItems.indexOfFirst { it.id == files.first().id }
+
+        for (index in firstFileIndex until firstFileIndex + files.size) {
+            chatItems[index] = (chatItems[index] as VisitorChatItem).withStatus(VisitorItemStatus.PREVIEW)
+        }
+    }
+
+    private fun mapMessageSent(message: VisitorMessage, state: State): State = mapNewMessage(ChatMessageInternal(message), state)
```
Comment (andrews-moc) [inline widgetssdk/src/main/java/com/glia/widgets/chat/ChatManager.kt:245]: Does it make sense to use `mapNewMessage()` directly, instead of a `mapMessageSent()` wrapper?

**Review by andrews-moc — COMMENTED** [review]
@DavDo You have done a tremendous scope of work for such a short period of time!

I reviewed as thoroughly, as I could. However, PR is so big, so I cannot be sure that I haven't missed anything, unfortunately. 😢

Such huge scope of changes should either be split into smaller pieces, iteration by iteration or we can only rely on manual and automated testing.

**Comment by DavDo** [comment]
> @DavDo You have done a tremendous scope of work for such a short period of time!
> 
> I reviewed as thoroughly, as I could. However, PR is so big, so I cannot be sure that I haven't missed anything, unfortunately. 😢
> 
> Such huge scope of changes should either be split into smaller pieces, iteration by iteration or we can only rely on manual and automated testing.

🙇‍♂️ I understand your point. 
The worst part of this kind of refactoring is that changing one small piece of business logic causes a change chain, and you just don't see a result until you finish all the related changes. In the meantime, the other part of business logic causes similar changes.
So, to split refactoring into small pieces of code, I think these small pieces must be atomic, but that is becoming clear (which parts of the change are atomic) only after finishing that big chain of changes.

I think manual tests are really important in this case because these changes have a big impact, and even with small PRs, it is unclear whether they work properly.

**Comment by andrews-moc** [comment]
@DavDo you need a help with fixing the unit tests?

**Comment by DavDo** [comment]
> @DavDo Do you need a help with fixing the unit tests?

No, Thanks. 🙇‍♂️ 
I'll create a separate PR soon for tests.

**Comment by andrews-moc** [comment]
> > @DavDo Do you need a help with fixing the unit tests?
> 
> No, Thanks. 🙇‍♂️ 
> I'll create a separate PR soon for tests.

Yeah, separate PR is a good idea. 👍
Also, I can test manually tomorrow, if you want.

**Comment by DavDo** [comment]
> > > @DavDo Do you need a help with fixing the unit tests?
> > 
> > 
> > No, Thanks. 🙇‍♂️
> > I'll create a separate PR soon for tests.
> 
> Yeah, separate PR is a good idea. 👍 Also, I can test manually tomorrow, if you want.

I have tested myself and also run ATs with these changes, but an extra pair of eyes would be very welcome.

File: widgetssdk/src/main/java/com/glia/widgets/chat/ChatManager.kt (hunk 2 above)
Comment (DavDo) [reply widgetssdk/src/main/java/com/glia/widgets/chat/ChatManager.kt]: done ✅

File: widgetssdk/src/main/java/com/glia/widgets/chat/ChatManager.kt (hunk 3 above)
Comment (DavDo) [reply widgetssdk/src/main/java/com/glia/widgets/chat/ChatManager.kt]: done ✅

File: widgetssdk/src/main/java/com/glia/widgets/chat/ChatContract.kt [hunk 4]
```diff
@@ -23,7 +23,7 @@ internal interface ChatContract {
         fun setView(view: View)
         fun getView(): View?
         fun onDestroy(retain: Boolean)
-        fun onMessageClicked(messageId: String)
+        fun onTapToRetryClicked(messageId: String)
```
Comment (gugalo) [inline widgetssdk/src/main/java/com/glia/widgets/chat/ChatContract.kt:26]: A bit lost if the 'Tap' and Clicked' in the same method name intentional and I just don't get it or just confusing naming.
If it is not intentional I would suggest `onMessageRetryClicked` to make it very clear that it is not just some retry but message retry

File: widgetssdk/src/main/java/com/glia/widgets/chat/ChatContract.kt [hunk 5]
```diff
@@ -57,6 +57,7 @@ internal interface ChatContract {
         fun onTakePhotoClicked()
         fun onImageCaptured(result: Boolean)
         fun onContentChosen(uri: Uri)
+        fun onLocalImageItemClick(attachment: LocalAttachment, view: android.view.View)
```
Comment (gugalo) [inline widgetssdk/src/main/java/com/glia/widgets/chat/ChatContract.kt:60]: 'Local' word for me has a bit of broad meaning, which required me to second gues what this method does. Could it be 'downloaded' or 'saved' instead?

File: widgetssdk/src/main/java/com/glia/widgets/chat/ChatContract.kt [hunk 6]
```diff
@@ -82,5 +83,6 @@ internal interface ChatContract {
         fun showToast(message: String, duration: Int = Toast.LENGTH_SHORT)
         fun dispatchImageCapture(uri: Uri)
         fun onFileDownload(attachmentFile: AttachmentFile)
+        fun navigateToPreview(attachmentFile: LocalAttachment, view: android.view.View)
```
Comment (gugalo) [inline widgetssdk/src/main/java/com/glia/widgets/chat/ChatContract.kt:86]: navigateToPreview -> navigateToFilePreview ?
I am a bit picky about naming )
But if you disagree with any of those then please share your thoughts I am happy to discuss and change my mind

File: widgetssdk/src/main/java/com/glia/widgets/chat/adapter/ChatAdapter.kt [hunk 7]
```diff
@@ -120,50 +116,28 @@ internal class ChatAdapter(
...
+                OperatorMessageViewHolder(ChatOperatorMessageLayoutBinding.inflate(inflater, parent, false), uiTheme)
 
-            MEDIA_UPGRADE_ITEM_TYPE -> {
-                MediaUpgradeStartedViewHolder(
-                    ChatMediaUpgradeLayoutBinding.inflate(inflater, parent, false),
-                    uiTheme
-                )
-            }
+            MEDIA_UPGRADE_ITEM_TYPE ->
+                MediaUpgradeStartedViewHolder(ChatMediaUpgradeLayoutBinding.inflate(inflater, parent, false), uiTheme)
```
Comment (gugalo) [inline widgetssdk/src/main/java/com/glia/widgets/chat/adapter/ChatAdapter.kt:135]: I wanted to ask for a while. Your PRs always have a lot of styling changes. Are these done manually or automatically? What are you using?
We could either agree to use the same tools by everyone in the team. Or agree to do those in separate PRs. I feel like more then third of this commit is styling changes.

**Review by gugalo — COMMENTED** [review]
You have done huge work which I always really appreciate in your PRs 👍 
And thank you for dividing it into two PRs ❤️ 
But still, you are killing me with the size of these changes. I just can't review all of this code with the required focus.

File: widgetssdk/src/main/java/com/glia/widgets/chat/ChatContract.kt (hunk 5 above)
Comment (DavDo) [reply widgetssdk/src/main/java/com/glia/widgets/chat/ChatContract.kt:60]: This is actually not downloaded or saved but a `local` file picked by the visitor.

File: widgetssdk/src/main/java/com/glia/widgets/chat/adapter/ChatAdapter.kt (hunk 7 above)
Comment (DavDo) [reply widgetssdk/src/main/java/com/glia/widgets/chat/adapter/ChatAdapter.kt:135]: Android Studio with the default profile: I think it takes formatting from our `.editorconfig` file, so it should be pretty much the same for each of us.

image

File: widgetssdk/src/main/java/com/glia/widgets/chat/ChatContract.kt (hunk 4 above)
Comment (DavDo) [reply widgetssdk/src/main/java/com/glia/widgets/chat/ChatContract.kt]: renamed - c64be489fee6ea4819df6498f1bd8e2ab9977e99

File: widgetssdk/src/main/java/com/glia/widgets/chat/ChatContract.kt (hunk 6 above)
Comment (DavDo) [reply widgetssdk/src/main/java/com/glia/widgets/chat/ChatContract.kt]: renamed - c64be489fee6ea4819df6498f1bd8e2ab9977e99

File: widgetssdk/src/main/java/com/glia/widgets/chat/adapter/ChatAdapter.kt (hunk 7 above)
Comment (andrews-moc) [reply widgetssdk/src/main/java/com/glia/widgets/chat/adapter/ChatAdapter.kt:135]: By the way, I also thought that part of the changes are related to formatting. And there is a chance, these kind of changes might be done in a separate PR. For example, do the coding part, commit, after that go through the changed files and apply automatic formatting, commit separately.

Comment (DavDo) [reply widgetssdk/src/main/java/com/glia/widgets/chat/adapter/ChatAdapter.kt:135]: This automatic formatting is part of the commit process, so it will not work after committing, and we have to manually walk through the files and apply formatting.

Comment (andrews-moc) [reply widgetssdk/src/main/java/com/glia/widgets/chat/adapter/ChatAdapter.kt:135]: Yes, that's why I do it manually. And sometimes miss to auto-format something. 🙂

**Review by andrews-moc — APPROVED** [review]
I have tested the functionality manually. Reported the issues to Davit. Thank you for the quick fixes, @DavDo.

 I'm approving the PR now, to unblock Davit's work (merging the unit tests PR, running acceptance tests).

To increase the confidence in quality, it might be good to do a team wide exploratory testing later.

File: widgetssdk/src/main/java/com/glia/widgets/chat/ChatManager.kt [hunk 8]
```diff
@@ -140,12 +145,19 @@ internal class ChatManager(
...
-            onChatAction(Action.MessageSent(it))
+        val payload = state.preEngagementChatItemIds.takeIf { it.isNotEmpty() }
+            ?.firstOrNull()
+            ?.let { state.messagePreviews[it] } ?: return
+
+        sendMessage(payload)
+    }
+
+    private fun sendMessage(payload: SendMessagePayload) {
+        println("*********************** -> ${payload.content}")
```
Comment (andrews-moc) [inline widgetssdk/src/main/java/com/glia/widgets/chat/ChatManager.kt:156]: @DavDo Don't forget to remove this line. 😉

Comment (DavDo) [reply widgetssdk/src/main/java/com/glia/widgets/chat/ChatManager.kt:156]: thanks
